libs/SalesforceSDK/src/com/salesforce/androidsdk/rest/RestClient.java

```diff
@@ -132,48 +136,71 @@
      * @param authTokenProvider
 	 */
 	public RestClient(ClientInfo clientInfo, String authToken, HttpAccess httpAccessor, AuthTokenProvider authTokenProvider) {
-		this(httpAccessor, new OAuthRefreshInterceptor(clientInfo, authToken, authTokenProvider));
-	}
-
-	public RestClient(HttpAccess httpAccessor, OAuthRefreshInterceptor httpInterceptor) {
+		this.clientInfo = clientInfo;
         this.httpAccessor = httpAccessor;
-		this.oAuthRefreshInterceptor = httpInterceptor;
+		this.authTokenProvider = authTokenProvider;
+		setOAuthRefreshInterceptor(authToken);
 		setOkHttpClientBuilder();
 		setOkHttpClient(null);
 	}
 
 	/**
-	 * Clear cache of org-id/user-id to OkHttpClient.Builder
+	 * Remove cached OkHttpClient.Builder, OkHttpClient and OAuthRefreshInterceptor for the given org id, user id
 	 */
-	public static void clearOkClientBuildersCache() {
-		OK_CLIENT_BUILDERS = null;
+	public synchronized static void clearCaches(String orgId, String userId) {
+		String cacheKey = computeCacheKey(orgId, userId);
```

Comment:
If we don't clean the caches, and a user logs out and then logs back in, he will use the old okhttpclient (with the old oauth refresh interceptor).
As a result the first request will fail since the auth header will use the old auth token.
In the case where refresh expires immediately, you end up being send back to login right away.